Adds odo registry command

cdrage opened the pull request:
<!--
Thank you for opening a PR! Here are some things you need to know before submitting:

1. Please read our developer guideline: https://github.com/redhat-developer/odo/wiki/Dev:-odo-Dev-Guidelines
2. Label this PR accordingly with the '/kind' line
3. Ensure you have written and ran the appropriate tests: https://github.com/redhat-developer/odo/wiki/Dev:-Writing-and-running-tests
4. Read how we approve and LGTM each PR: https://github.com/redhat-developer/odo/wiki/Pull-Requests:-Review-guideline

Documentation:

If you are pushing a change to documentation, please read: https://github.com/redhat-developer/odo/wiki/Documentation:-Contributing
-->

**What type of PR is this:**

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind tests
/kind documentation

Feel free to use other [labels](https://github.com/redhat-developer/odo/labels) as needed. However one of the above labels must be present or the PR will not be reviewed. This instruction is for reviewers as well.
-->
/kind feature

**What does this PR do / why we need it:**

This adds the `odo registry` command, to output the details of what's
in the registry.

It also implements the flags:
- `--filter`
- `--devfile`
- `--details`

This can be tested by typing in the `odo registry` command after adding
a registry (you should already have one by default)

**Which issue(s) this PR fixes:**
<!--
Specifying the issue will automatically close it when this PR is merged
-->

part of https://github.com/redhat-developer/odo/issues/5524

**PR acceptance criteria:**

- [X] Unit test

- [X] Integration test

- [X] Documentation

**How to test changes / Special notes to the reviewer:**

`odo registry`

dharmit commented on pkg/odo/cli/registry/registry.go:
```diff
@@ -0,0 +1,213 @@
+
+	// Flags
+	listCmd.Flags().StringVar(&o.filterFlag, "filter", "", "Filter based on the name of the component")
+	listCmd.Flags().StringVar(&o.devfileFlag, "devfile", "", "Only show components from the specific Devfile stack")
```
Playing with this, I typed individual Devfile stack name like `python` and `python-django`, but that failed. Passed the name of the registry and it worked. So, IMO, this should be reworded to below.:
```suggestion
	listCmd.Flags().StringVar(&o.devfileFlag, "devfile", "", "Only show components from the specific Devfile registry")
```

dharmit commented on pkg/odo/cli/registry/registry.go:
```diff
@@ -0,0 +1,213 @@
+	t := table.NewWriter()
+	rowConfigAutoMerge := table.RowConfig{AutoMerge: true}
+
+	// If passed int he details flag, we'll be using a different style of table output.
```
```suggestion
	// If passed in the details flag, we'll be using a different style of table output.
```

dharmit commented on pkg/odo/cli/registry/registry.go:
```diff
@@ -0,0 +1,213 @@
+
+	// If we pass in the details flag, we will manipulate the table as well as create some
+	if o.detailsFlag {
+		t.AppendHeader(table.Row{"NAME / VERSION", "REGISTRY / URL", "LANGUAGE / TAGS", "DEV", "DEPLOY", "DEBUG", "DESCRIPTION", "STARTER PROJ."}, rowConfigAutoMerge)
```
IMO, keeping the `REGISTRY/URL` column a bit later might be better. Saying this because in a real registry, there are going to be a bunch of stacks and the registry info might not even be immediately useful to users not working with multiple registries. 

In the screenshot, you can see that I have only one registry and that is occupying a good amount of space which causes info about starter projects difficult to get.
![Screenshot from 2022-04-18 15-45-47](https://user-images.githubusercontent.com/713521/163795161-0e94d063-1fb9-4cbf-8331-97eae399d174.png)

kadel commented:
That tabular output doesn't work at all :-/ 
![image](https://user-images.githubusercontent.com/57206/164034514-61360701-56e8-4930-bdeb-b949a7b32160.png)

kadel commented:
> Fixes #5524
`--devfile` flag doesn't seems to work, and there is no `--devfile-registry` flag.

cdrage commented:
> That tabular output doesn't work at all :-/ ![image](https://user-images.githubusercontent.com/57206/164034514-61360701-56e8-4930-bdeb-b949a7b32160.png)

Argh! That's unfortunate. I'll play around and probably switch to the "list" output, as much as the tabbed output looks better.

cdrage commented:
/hold

Have to wait for https://github.com/redhat-developer/odo/pull/5664 to be merged in so we have support for detecting if a Devfile is using DEBUG / DEPLOY / DEV commands.

cdrage commented:
Updated PR to now list components with `--details` as a list.

I've also added `--devfile` and `--devfile-registry` so you can specify 1 specific Devfile, as well as what registry you wish to query.

```sh
Name: python-django
Display Name: Django
Registry: DefaultDevfileRegistry
Registry URL: https://registry.devfile.io
Version: 1.0.0
Description: Python3.7 with Django
Tags: Python, pip, Django
Project Type: django
Language: python
Starter Projects:
  - django-example
Supported odo features:
  - dev: Y
  - deploy: Y
  - debug: Y
```

cdrage commented:
Example:

```sh
go/src/odo  add-registry-command ✗                                                                                                                                                                                                                                                                                                                                28m ⚑
▶ odo registry --devfile-registry Test --devfile python
 NAME    REGISTRY  DESCRIPTION
 python  Test      Python Stack with Python 3.7
 ```

kadel commented on pkg/odo/cli/registry/registry.go:
```diff
@@ -0,0 +1,214 @@
+				devfileComponent.ProjectType,
+				devfileComponent.Language,
+				strings.Join(devfileComponent.StarterProjects, "\n  - "),
+				// TODO, showing dev / deploy / debug NOT yet implemented
```
If there are things that are not implemented yet, and you plan to add it in this PR, please mark PR as work in progress.

dharmit replied:
@cdrage this part is still commented. Is it going to be handled in this PR or a subsequent one?

kadel commented:
I haven't noticed any e2e or unit tests :-/

cdrage commented:
Unit and integration tests have now been added 👍

cdrage commented:
/retest

cdrage commented:
Ready for review!

kadel commented:
if I specify non-existing`--devfile` it silently exits with no error :/ even the exit code is `0`

```
▶ odo registry --details --devfile non

▶ echo $?
0
```

kadel commented:
`--filter` doesn't seem to search in Description.

```
▶ odo registry --filter OpenJDK
 NAME  REGISTRY  DESCRIPTION

▶ odo registry | grep  OpenJDK
 java-maven                    DefaultDevfileRegistry  Upstream Maven and OpenJDK 11
 java-maven                    StageRegistry           Upstream Maven and OpenJDK 11
```

cdrage commented:
Hey @kadel !

Thanks for the review, I've updated it so it now exists with an error if there is nothing found, as well as updated the `--filter ` so that it includes searching through the description:

```sh
go/src/odo  add-registry-command ✗                                                                                                                                                                                                                                                                                                                               1[14/152]
▶ odo registry --filter kalsjdfklasjdf
 ✗  There are no Devfiles available to show

go/src/odo  add-registry-command ✗                                                                                                                                                                                                                                                                                                                              1d0h ⚑  ⍉
▶ odo registry --filter Java application --devfile-registry Test
 NAME                          REGISTRY  DESCRIPTION
 java-openliberty              Test      Java application Maven-built stack using...
 java-openliberty-gradle       Test      Java application Gradle-built stack usin...
 java-quarkus                  Test      Quarkus with Java
 java-springboot               Test      Spring Boot® using Java
 java-vertx                    Test      Upstream Vert.x using Java
 java-websphereliberty         Test      Java application Maven-built stack using...
 java-websphereliberty-gradle  Test      Java application Gradle-built stack usin...
 java-wildfly-bootable-jar     Test      Java stack with WildFly in bootable Jar ...
```

valaparthvi commented on pkg/odo/cli/registry/registry.go:
```diff
@@ -0,0 +1,226 @@
+
+	}
+
+	// Exit with an error if there are no components to show, so we don't render the table / continue
+	if len(devfiles) == 0 {
+		log.Error("There are no Devfiles available to show")
+		return
+	}
```
Can you move this before the for loop? I know for loop is not going to be bothered by 0 length, but it just makes sense to me to check the length and return early on if required.

cdrage replied:
I put it after the for loop because the for loop ADDS to `devfiles` variable.

valaparthvi commented on pkg/registry/registry_test.go:
```diff
@@ -87,6 +88,153 @@ OdoSettings:
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			prefClient, _ := preference.NewClient()
```
Can you use Mock client instead? See https://github.com/redhat-developer/odo/blob/main/pkg/init/init_test.go#L65 for reference.

cdrage replied:
Didn't need to since we're not modifying the clients (clients are both "blank"), so it was cleaner to implement it this way, instead of via fields.

What do you think @feloy ?

feloy replied:
By mocking ths Preference client, you could mock the methods on this client the tested function is using only, without needing to create the underlying file. This way, you should not rely on the implementation of the preference package

cdrage replied:
@feloy Unfortunately https://github.com/redhat-developer/odo/blob/main/pkg/registry/registry.go#L175 has way too many HTTP requests and because it's not part of the preference or registry client, I'd be unable to mock it without refactoring that side of code (which is outside the scope of this PR).

Could we keep these tests (which are based on the `odo dev` unit tests) and revisit it at a later date to refactor?

At the moment, they pass, they work well.

feloy replied:
@cdrage I don't think it is related. You can see on this commit how to mock the preference client to avoid using a preference file:
https://github.com/feloy/odo/commit/6551ebe2d74281b87b0a4b50f76ba78168da1713

cdrage replied:
Ohhhh! I thought you were asking me to also somehow mock the http server... which would of been a lot.

I'll add the preference mock

valaparthvi reviewed (changes requested):
Everything worked great for me, but I just have a few comments.

cdrage commented on pkg/odo/cli/registry/registry.go:
```diff
@@ -0,0 +1,226 @@
+
+	t.SetStyle(table.Style{
+		Box: table.BoxStyle{
+			PaddingLeft:  " ",
+			PaddingRight: " ",
```
Unfortunately I've tried this with more padding, but found that it would "wrap" the terminal if there's not enough space :( so we'll have to keep it as small as possible.

feloy commented on pkg/odo/cli/registry/registry.go:
```diff
@@ -0,0 +1,226 @@
+%[1]s --filter nodejs --devfile-registry DefaultDevfileRegistry
+
+# Show more details
+%[1]s --details`
```
Could you add the example with  `--details --devfile <name> --devfile-registry <registry>` to get the details of a named devfile?

feloy commented on pkg/odo/cli/registry/registry.go:
```diff
@@ -0,0 +1,226 @@
+	clientset *clientset.Clientset
+
+	// List of known devfiles
+	DevfileList registry.DevfileStackList
```
You can make it non visible (`devfileList`)

feloy commented on pkg/odo/cli/registry/registry.go:
```diff
@@ -0,0 +1,226 @@
+	t.SetOutputMirror(log.GetStdout())
+
+	t.AppendHeader(table.Row{"NAME", "REGISTRY", "DESCRIPTION"})
+	t.SortBy([]table.SortBy{
```
I think we want to order by registry in order of appareance in the config (the last in the list appearing first):

```
$ cat ~/.odo/preference.yaml
kind: Preference
apiversion: odo.dev/v1alpha1
OdoSettings:
  RegistryList:
  - Name: DefaultDevfileRegistry
    URL: https://registry.devfile.io
    secure: false
  - Name: MyRegistry
    URL: https://registry.stage.devfile.io
    secure: false
  Ephemeral: false
  ConsentTelemetry: true


$ odo registry
 NAME                          REGISTRY                DESCRIPTION                                 
 dotnet50                      DefaultDevfileRegistry  Stack with .NET 5.0                         
 dotnet50                      MyRegistry              Stack with .NET 5.0                         
 dotnet60                      DefaultDevfileRegistry  Stack with .NET 6.0                         
 dotnet60                      MyRegistry              Stack with .NET 6.0                         
 dotnetcore31                  DefaultDevfileRegistry  Stack with .NET Core 3.1                    
 dotnetcore31                  MyRegistry              Stack with .NET Core 3.1                    
```

should be

```
$ odo registry
 NAME                          REGISTRY                DESCRIPTION                                 
 dotnet50                      MyRegistry              Stack with .NET 5.0                         
 dotnet50                      DefaultDevfileRegistry  Stack with .NET 5.0                         
 dotnet60                      MyRegistry              Stack with .NET 6.0                         
 dotnet60                      DefaultDevfileRegistry  Stack with .NET 6.0                         
 dotnetcore31                  MyRegistry              Stack with .NET Core 3.1                    
 dotnetcore31                  DefaultDevfileRegistry  Stack with .NET Core 3.1                    
```

feloy commented on pkg/odo/cli/registry/registry.go:
```diff
@@ -0,0 +1,226 @@
+
+		// If the user has specified a filter with variable o.filterFlag, then only show the components
+		// containing that specific string.
+		if o.filterFlag != "" {
```
Filtering should happen in business level IMO

feloy commented on pkg/odo/cli/registry/registry.go:
```diff
@@ -0,0 +1,226 @@
+		}
+
+		// If the user passed in --devfile-registry <REGISTRY-NAME>, then only show the components from that Devfile stack
+		if o.registryFlag != "" {
```
idem as filtering

feloy commented on pkg/odo/cli/registry/registry.go:
```diff
@@ -0,0 +1,226 @@
+		}
+
+		// If the user passed in --devfile <NAME> only show that specific component matching that name
+		if o.devfileFlag != "" {
```
idem as filtering

feloy commented:
> Well.. that's unfortunate. With the registry schema, we can't get the commands used within the devfile, and in order to implement the "Odo supported features:" part of odo registry, we have to make an upstream change.
> 
> @feloy implementation is based upon parsing devfile.yaml / object.
> 
> This means that in order to show it within odo registry would have to download each devfile.yaml, parse it, and then show it...
> 
> We can either:
> 
> 1. Implement odo registry and open up an issue upstream (I've opened it up here: [Implement a commands variable in the registry schema for easier group ID retrieval. devfile/api#831](https://github.com/devfile/api/issues/831)). Merge this PR as-is without showing "Odo supported features"
> 2. Download and parse each devfile.yaml whenever you run --details for each devfile component listed in order to show "Odo supported features".

I think we discussed and selected the second solution: Download and parse each devfile.yaml whenever you run --details for each devfile component listed in order to show "Odo supported features".

cdrage commented:
> > Well.. that's unfortunate. With the registry schema, we can't get the commands used within the devfile, and in order to implement the "Odo supported features:" part of odo registry, we have to make an upstream change.
> > @feloy implementation is based upon parsing devfile.yaml / object.
> > This means that in order to show it within odo registry would have to download each devfile.yaml, parse it, and then show it...
> > We can either:
> > 
> > 1. Implement odo registry and open up an issue upstream (I've opened it up here: [Implement a commands variable in the registry schema for easier group ID retrieval. devfile/api#831](https://github.com/devfile/api/issues/831)). Merge this PR as-is without showing "Odo supported features"
> > 2. Download and parse each devfile.yaml whenever you run --details for each devfile component listed in order to show "Odo supported features".
> 
> I think we discussed and selected the second solution: Download and parse each devfile.yaml whenever you run --details for each devfile component listed in order to show "Odo supported features".

Okay, that'll probably add a few days to the implementation then. I've opened up the upstream feature request anyways, but i'll implement it in this PR.

cdrage commented:
/retest

cdrage commented:
/retest

cdrage commented:
I/O timeout

Retesting

/retest

cdrage commented:
Tests added! Ready for a review 👍

feloy commented on pkg/odo/cli/registry/registry.go:
```diff
@@ -0,0 +1,241 @@
+	devfiles := []registry.DevfileStack{}
+
+	// Filter through all the devfile components per the filters / parameters passed in.
+	for _, devfileComponent := range DevfileList {
+
+		// If the user has specified a filter with variable o.filterFlag, then only show the components
+		// containing that specific string.
+		if o.filterFlag != "" {
+			if !strings.Contains(devfileComponent.Name, o.filterFlag) && !strings.Contains(devfileComponent.Description, o.filterFlag) {
+				continue
+			}
+		}
+
+		// If the user passed in --devfile <NAME> only show that specific component matching that name
+		if o.devfileFlag != "" {
+			if devfileComponent.Name != o.devfileFlag {
+				continue
+			}
+		}
+		devfiles = append(devfiles, devfileComponent)
+	}
```
I think you can remove this code now the filtering is done elsewhere

feloy commented on pkg/odo/cli/registry/registry.go:
```diff
@@ -0,0 +1,241 @@
+
+			// If details are being shown, we must download the Devfile from the registry
+			// parse it, in order to retrieve command information such as if debug, dev, deploy and test is enabled.
+			devfileData, err := o.clientset.RegistryClient.RetrieveDevfileDataFromRegistry(devfileComponent.Registry.Name, devfileComponent.Name)
```
We will also need this data for the JSON output, please move this code to the business layer

cdrage replied:
Have to keep it in this layer.. because this is only triggered with `--details`. If we had this in the business layer, we would http get every single devfile which is what we don't want.

Or I could add it as a boolean of true/false for the function if we require details or not? What do you think?

feloy replied:
Yes, you could pass an option to the business layer function to indicate if you want or not the details

cdrage commented:
@feloy I've gone ahead and moved that functionality down to the business layer as well as improved tests.

cdrage commented:
@feloy Updated the code to add the sorting logic 👍

feloy commented:
@cdrage the OpenShift test is failing on one test. Could you check if your PR is completely rebased on main, and re-push the latest commit to restart the tests?

cdrage commented:
> @cdrage the OpenShift test is failing on one test. Could you check if your PR is completely rebased on main, and re-push the latest commit to restart the tests?

Yup, repushing now

feloy commented:
/lgtm
/approve
